bpf/flows.c

```diff
@@ -155,29 +155,7 @@ static inline int flow_monitor(struct __sk_buff *skb, u8 direction) {
             update_existing_flow(aggregate_flow, &pkt, len, filter_sampling);
         } else if (skb->ifindex != 0) {
             // Only add info that we've seen this interface
-            additional_metrics *extra_metrics =
-                (additional_metrics *)bpf_map_lookup_elem(&additional_flow_metrics, &id);
-            if (extra_metrics != NULL) {
-                add_observed_intf(extra_metrics, skb->ifindex, direction);
-            } else {
-                additional_metrics new_metrics = {
-                    .eth_protocol = eth_protocol,
-                    .start_mono_time_ts = pkt.current_ts,
-                    .end_mono_time_ts = pkt.current_ts,
-                };
-                add_observed_intf(&new_metrics, skb->ifindex, direction);
-                long ret =
-                    bpf_map_update_elem(&additional_flow_metrics, &id, &new_metrics, BPF_NOEXIST);
-                if (ret == -EEXIST) {
-                    extra_metrics =
-                        (additional_metrics *)bpf_map_lookup_elem(&additional_flow_metrics, &id);
-                    if (extra_metrics != NULL) {
-                        add_observed_intf(extra_metrics, skb->ifindex, direction);
-                    }
-                } else if (ret != 0 && trace_messages) {
-                    bpf_printk("error creating new observed_intf: %d\n", ret);
-                }
-            }
+            add_observed_intf(aggregate_flow, skb->ifindex, direction);
```

Comment:
need to acquire lock here

Reply:
make interface update part of `update_existing_flow()`

Reply:
done, but I still take the lock in add_observed_intf for 2 reasons:
- we can't call a function while having a lock (so for instance the call to `increase_counter` isn't allowed)
- we can't read map value while having the lock, so all the for-loop to check for existing intf+direction isn't allowed during locking
In other words we need to keep lock as short time as possible